emubd/lfs_emubd.c

```diff
@@ -190,7 +190,7 @@ int lfs_emubd_erase(const struct lfs_config *cfg, lfs_block_t block) {
     }
 
     if (!err && S_ISREG(st.st_mode) && (S_IWUSR & st.st_mode)) {
-        int err = unlink(emu->path);
+        err = unlink(emu->path);
```

Comment:
This is a bug as it overrides the value of `err` used on line 199 (6 lines below this). It's passing because of undefined behaviour below when we use the `st` variable after `stat` failed.

Reply:
I don't follow.

    int err = stat(emu->path, &st);
    if (err && errno != ENOENT) {
        return -errno;
    }

    if (!err && S_ISREG(st.st_mode) && (S_IWUSR & st.st_mode)) {
/* err == 0 here, no harm to set it */
        err = unlink(emu->path);
        if (err) {
            return -errno;
        }
/* ditto */
    }
/* either stat() gave 0 (and left errno at it's value before stat()) or !0 and errno == ENOENT */
/* I.e. the first condition below is superfluous but we still should check if the open for write would succeed, i suppose */
    if (err ||  (S_ISREG(st.st_mode) && (S_IWUSR & st.st_mode))) {

Now this whole thing is heavier than needed (either blindly unlink() or cowardly access(path, W_OK) before unlinking but then again, it's just the emu thing. What am i missing? :)

Reply:
```c
int lfs_emubd_erase(const struct lfs_config *cfg, lfs_block_t block) {
    lfs_emubd_t *emu = cfg->context;

    // Check if erase is valid
    assert(block < cfg->block_count);

    // Erase the block
    int err = snprintf(emu->child, LFS_NAME_MAX, "%x", block);
    assert (err > 0 && err < LFS_NAME_MAX); /* truncated or other error */

    if (access(emu->path, F_OK) == 0) {
        FILE *f = fopen(emu->path, "w");
	if (f && fclose(f) == 0) {
	    emu->stats.erase_count += 1;
	    return 0;
	}
    }
    return -errno;
}
```
```shell
$ ../busybox/scripts/bloat-o-meter gcc7.a gcc7-silence.a 
function                                             old     new   delta
lfs_emubd_erase                                      203     128     -75
------------------------------------------------------------------------------
(add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-75)             Total: -75 bytes
$
```
Just a thought..

Reply:
Oh, you're right, my bad. Though these are the sort of cases -Wshadow makes me worried about.

Either way I think your restructure of lfs_emubd_erase is a good idea. Although counterintuitively, lfs_emubd_erase should return 0 even if access fails. The tests remove write permission to simulate a block that has frozen data.

Also out of curiousity, is `fopen(path, "w"), fclose(f)` prefered over `truncate(path)` in this situation?

Reply:
I've saved that away as https://github.com/aldot/littlefs/commit/1012068bcf13b0de6462f80ca081a8a20388229d now, fwiw.
Note that i didn't mean to spend any time whatsoever on emubd :P

Reply:
well open(path,O_CREAT|O_TRUNC|O_NOFOLLOW|O_WRONLY,0644)/close might be an even better choice, sure.

Reply:
Haha that's fair. Welp, it's not an issue with this pr like I thought it was.

Reply:
but Re: "Although counterintuitively, lfs_emubd_erase should return 0 even if access fails. The tests remove write permission to simulate a block that has frozen data."

but that isn't the case currently, is it?
format fails with read-only block 0 when erase() is just
```C
    int fd = open(emu->path, O_CREAT|O_TRUNC|O_NOFOLLOW|O_WRONLY, 0644);        
    if (fd >= 0 && close(fd) == 0) {                                            
        emu->stats.erase_count += 1;                                            
        return 0;                                                               
    }                                                                           
    return -errno;
```
I thought that an erase of a "bad" block (i.e. RO) have to get us a "good" block (i.e. WR) block.
At least that's the effect of unlink() in your current code.
What am i missing?

Reply:
But generally speaking, switching just erase() to fd-based ops doesn't make sense.
If we don't question the usefulness of wasting time on emubd in the first place for a second, all of it should be fd-based and not FILE-based. The latter is very, very complicated -- think about all that buffering and behind the scenes maintenance -- and doesn't make sense for something like emubd i'd say. No?

Reply:
It's not the prettiest code, but that's what the if statement around unlink is for. If the file is not writeable it would not try to unlink.
``` c
if (!err && S_ISREG(st.st_mode) && (S_IWUSR & st.st_mode)) {
    err = unlink(emu->path);
    ...
```

The purpose of using unlink was to indicate to the `lfs_emubd_prog` function that the block was actually erased, but it might be possible to replace that logic with a check against the file size.

https://github.com/aldot/littlefs/blob/962b0ea5ad941925737be17c0d69404d630b6a1d/emubd/lfs_emubd.c#L140-L146